internal/pkg/addons/cloudformation.go

```diff
@@ -103,19 +104,39 @@ func (t *cfnTemplate) mergeTransform(transform yaml.Node) error {
 // mergeResources updates t's Resources with additional resources.
 // If a resource already exists with a different value, returns errResourceAlreadyExists.
 func (t *cfnTemplate) mergeResources(resources yaml.Node) error {
-	if err := mergeSingleLevelMaps(&t.Resources, &resources); err != nil {
-		return wrapKeyAlreadyExistsErr(resourcesSection, err)
-	}
-	return nil
+	return mergeSingleLevelMaps(&t.Resources, &resources)
 }
 
 // mergeOutputs updates t's Outputs with additional outputs.
 // If an output already exists with a different value, returns errOutputAlreadyExists.
 func (t *cfnTemplate) mergeOutputs(outputs yaml.Node) error {
-	if err := mergeSingleLevelMaps(&t.Outputs, &outputs); err != nil {
-		return wrapKeyAlreadyExistsErr(outputsSection, err)
+	return mergeSingleLevelMaps(&t.Outputs, &outputs)
+}
+
+// assignNewNodesTo associates every new node added to the template t with the tplName.
+func (t *cfnTemplate) assignNewNodesTo(tplName string) {
+	var assign func(node *yaml.Node)
+	assign = func(node *yaml.Node) {
+		if node == nil {
+			return
+		}
+		if _, ok := t.templateNameFor[node]; ok {
+			// node is already associated with a previous template.
+			return
+		}
+		t.templateNameFor[node] = tplName
+		for _, c := range node.Content {
+			assign(c)
+		}
 	}
-	return nil
+
+	assign(&t.Metadata)
```

Comment:
I was thinking maybe we can simplify by:

1) starting from the entry node for this template and do the bst (assuming we don't think it is valid to have "nodeA" defined in both `Metadata` of templateA and `Parameters` of templateB) 

or

2) make it a new field like `t.Content map[string]*node` instead to have `t.Parameters`, `t.Mappings` etc. By doing this we can change it to two-line for loop for this `assignNewNodesTo` and also we can remove `mergeMetadata` etc. to have a generic function with handler as a param.

What do you think of it 🤔

Reply:
I used reflection to call `assign` on all the fields that represent CFN sections. Take a look at it and let me know what you think.

This way we don't have to update this function if we have to add new section

Reply:
Makes sense to me :+1: Thank you for the extra efforts to make this change!